### PR TITLE
v0.23.x: Add support for HTTP Basic authentication to HTTP and WebSocket RPC clients

### PR DESCRIPTION
**Pull request by romac**
Closes: #1169

---

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Added entry in `.changelog/`

### CODE REVIEW

File: rpc/src/rpc_url.rs [hunk 1]
```diff
@@ -98,15 +99,22 @@ impl Url {
...
     /// Get the password associated with this URL, if any.
     pub fn password(&self) -> Option<&str> {
         self.inner.password()
     }
 
+    /// Get the authority associated with this URL, if any.
+    /// The authority is the username and password separated by a colon.
+    pub fn authority(&self) -> Option<String> {
+        self.username()
+            .map(|user| format!("{user}:{}", self.password().unwrap_or_default()))
```
Comment (thanethomson) [inline rpc/src/rpc_url.rs:115]: Seems like Rust nightly doesn't like this?

File: rpc/src/rpc_url.rs [hunk 2]
```diff
@@ -179,7 +187,7 @@ mod test {
                     host: "127.0.0.1".to_string(),
```
Comment (thanethomson) [inline rpc/src/rpc_url.rs:187]: Could you please add a couple more test cases to `SUPPORTED_URLS` here? e.g. one with just a username, one with username and password.

**Review by thanethomson — COMMENTED** [review]
On first pass this mostly looks good to me! Just a couple of minor things.

File: rpc/src/rpc_url.rs (hunk 1 above)
Comment (romac) [reply rpc/src/rpc_url.rs]: Fixed

**Comment by romac** [comment]
Should be good to go :)

**Review by thanethomson — APPROVED** [review]
Excellent, thanks @romac!

**Pull request merged**